publish-mqtt/src/main.rs

```diff
@@ -167,35 +199,44 @@ async fn requests(mut homie: HomieDevice) -> Result<(), Box<dyn Error>> {
 
     homie.ready().await?;
 
-    let mut sensors_connected: Vec<BluetoothDevice> = vec![];
+    let mut sensors_connected: Vec<Sensor> = vec![];
 
     loop {
         println!("{} sensors in queue to connect.", sensors_to_connect.len());
         // Try to connect to a sensor.
-        if let Some(sensor) = sensors_to_connect.pop_front() {
-            let mac_address = sensor.get_address()?;
-            let name = sensor_names.get(&mac_address).unwrap_or(&mac_address);
-            println!("Trying to connect to {}", name);
-            match connect_start_sensor(
-                bt_session,
-                &mut homie,
-                &sensor_names,
-                properties.to_vec(),
-                &sensor,
-            )
-            .await
+        if let Some(mut sensor) = sensors_to_connect.pop_front() {
+            println!("Trying to connect to {}", sensor.name());
+            match connect_start_sensor(bt_session, &mut homie, properties.to_vec(), &mut sensor)
+                .await
             {
                 Err(e) => {
-                    println!("Failed to connect to {}: {:?}", name, e);
+                    println!("Failed to connect to {}: {:?}", sensor.name(), e);
                     sensors_to_connect.push_back(sensor);
                 }
                 Ok(()) => {
-                    println!("Connected to {} and started notifications", name);
+                    println!("Connected to {} and started notifications", sensor.name());
                     sensors_connected.push(sensor);
                 }
             }
         }
 
+        // If a sensor hasn't sent any updates in a while, disconnect it and add it back to the
+        // connect queue.
+        let now = Instant::now();
+        if let Some(sensor_index) = sensors_connected
+            .iter()
+            .position(|s| now - s.last_update_timestamp > UPDATE_TIMEOUT)
+        {
+            let sensor = sensors_connected.remove(sensor_index);
+            println!(
+                "No update from {} for {:?}, reconnecting",
+                sensor.name,
+                now - sensor.last_update_timestamp
+            );
+            homie.remove_node(&sensor.node_id()).await?;
+            sensors_to_connect.push_back(sensor);
```

Comment:
Should we also disconnect from it via bluetooth? Would this increase our chances of getting properly connected when it comes up in the queue?

Reply:
I'm not entirely sure. From when I was running it before it seemed to work without disconnecting.